src/Middleware/WebSockets/src/ServerWebSocket.cs

```diff
@@ -0,0 +1,80 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.WebSockets;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.WebSockets;
+
+/// <summary>
+/// Used in ASP.NET Core to wrap a WebSocket with its associated HttpContext so that when the WebSocket is aborted
+/// the underlying HttpContext is aborted. All other methods are delegated to the underlying WebSocket.
+/// </summary>
+internal sealed class ServerWebSocket : WebSocket
+{
+    private readonly WebSocket _wrappedSocket;
+    private readonly HttpContext _context;
+
+    internal ServerWebSocket(WebSocket wrappedSocket, HttpContext context)
+    {
+        ArgumentNullException.ThrowIfNull(wrappedSocket);
+        ArgumentNullException.ThrowIfNull(context);
+
+        _wrappedSocket = wrappedSocket;
+        _context = context;
+    }
+
+    public override WebSocketCloseStatus? CloseStatus => _wrappedSocket.CloseStatus;
+
+    public override string? CloseStatusDescription => _wrappedSocket.CloseStatusDescription;
+
+    public override WebSocketState State => _wrappedSocket.State;
+
+    public override string? SubProtocol => _wrappedSocket.SubProtocol;
+
+    public override void Abort()
```

Comment:
I know we use this pattern everywhere (override and add code), but what happens if someone add a virtual overload to the base type?  How would we find out so we could hook that one too?

Reply:
We've sometimes written tests that reflect on the type and check for anything missing. Probably not worth the effort, I don't think anything virtual has been added to this type since it was first written for .NET 4.5.

Reply:
> I know we use this pattern everywhere (override and add code), but what happens if someone add a virtual overload to the base type? How would we find out so we could hook that one too?

This was a big issue for streams and we built an analyzer to help detect this. WebSocket isn't pervasive enough to warrant it but cc @stephentoub.